### PR TITLE
[TOREVIEW] Added a MetadataParser util class for pages metadata

### PR DESCRIPTION
**Pull request by MazieroMarco**
The MetadataParser class can extract the metadata contained in a md file and returns it to the markdown-html converter to add it in the html header. The metadata, if extracted successfully, will be removed from the given source file.

Linked issue : #28

### CODE REVIEW

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            throw new RuntimeException("Incorrect " + head +  " line");
+    }
+
+    public static Metadata exctractMetadata(File mdFile) throws RuntimeException {
+        // Checks file validity
+        if (mdFile.isDirectory())
+            throw new RuntimeException("File cannot be a directory");
+
+        String filename = mdFile.getName();
+        if (!filename.substring(filename.lastIndexOf(".") + 1).equals("md"))
```
Comment (LeonardBesseau) [inline src/main/java/ch/heig_vd/app/utils/MetadataParser.java:21]: If we have FilenameUtils included why not use it ?
http://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FilenameUtils.html#getExtension%28java.lang.String%29

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        // Local vars
+        int nbMetaLines = 4;
+        String[] metaLines = new String[nbMetaLines];
+        Metadata outputMeta = new Metadata();
+
+        // Reads the file
+        BufferedReader reader;
+        try {
+            reader = new BufferedReader(new FileReader(mdFile));
+            reader.mark(1);
```
Comment (LeonardBesseau) [inline src/main/java/ch/heig_vd/app/utils/MetadataParser.java:33]: Why use mark ? Do we read these lines multiples time ?

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java (hunk 1 above)
Comment (MazieroMarco) [reply src/main/java/ch/heig_vd/app/utils/MetadataParser.java]: Updated. FilenameUtils is indeed more practical, didn't know it existed :)

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            mdFile.delete();
+            tmpFile.renameTo(new File(filename));
+
+            // Returns the parsed metadata
+            return outputMeta;
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        return outputMeta;
```
Comment (LeonardBesseau) [inline src/main/java/ch/heig_vd/app/utils/MetadataParser.java:77]: Why the return here ? it should not be accessed

**Review by LeonardBesseau — COMMENTED** [review]
Will check test later

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java (hunk 3 above)
Comment (MazieroMarco) [reply src/main/java/ch/heig_vd/app/utils/MetadataParser.java:75]: Fixed.

File: src/main/java/ch/heig_vd/app/utils/Metadata.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package ch.heig_vd.app.utils;
+
+public class Metadata {
+    public String title;
```
Comment (LeonardBesseau) [inline src/main/java/ch/heig_vd/app/utils/Metadata.java:4]: Do we really make them public ?

Comment (LeonardBesseau) [reply src/main/java/ch/heig_vd/app/utils/Metadata.java:4]: If not does the class need to be public ?

File: src/test/java/ch/heig_vd/app/utils/MetadataParserTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void givenFileShouldNotBeDirectory() {
+        File dir = new File("./");
+        MetadataParser.exctractMetadata(dir);
+    }
+
```
Comment (LeonardBesseau) [inline src/test/java/ch/heig_vd/app/utils/MetadataParserTest.java:50]: Maybe add test to check if metadata is malformed

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                throw new RuntimeException("Missing end of metadata line");
+
+            // Parses the extracted lines
+            outputMeta.title = parseMetaLine("titre:", metaLines[0]);
+            outputMeta.author = parseMetaLine("auteur:", metaLines[1]);
+            outputMeta.date = parseMetaLine("date:", metaLines[2]);
+
+            // creates temp file to write in
+            File tmpFile = new File(FilenameUtils.getPath(mdFile.getPath()) + "/tmp" + mdFile.getName());
+            FileWriter writer = new FileWriter(tmpFile);
```
Comment (LeonardBesseau) [inline src/main/java/ch/heig_vd/app/utils/MetadataParser.java:53]: Do we use a BufferedWriter here ?

**Comment by MazieroMarco** [comment]
Update : The MetadataParser needs to be separated into two distinct methods:
- extractMetadata -> reads the given file and extracts all the metadata inside
- extractContent -> reads the given file and extracts all the markdown content

File: src/main/java/ch/heig_vd/app/utils/PageParser.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    private static String parseMetaLine(String head, String line) {
+        if (line.substring(0, head.length()).equals(head))
+            return line.substring(head.length());
+        else
+            throw new RuntimeException("Incorrect " + head +  " line");
+    }
+
+    public static Metadata exctractMetadata(File mdFile) throws RuntimeException {
```
Comment (LeonardBesseau) [inline src/main/java/ch/heig_vd/app/utils/PageParser.java:24]: Typo in the name of the function

**Review by LeonardBesseau — CHANGES_REQUESTED** [review]
Just fix the typo and it is all good

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java (hunk 1 above)
Comment (LeonardBesseau) [reply src/main/java/ch/heig_vd/app/utils/MetadataParser.java]: Yeah. It is something to know.

File: src/main/java/ch/heig_vd/app/utils/MetadataParser.java (hunk 2 above)
Comment (LeonardBesseau) [reply src/main/java/ch/heig_vd/app/utils/MetadataParser.java]: Is corrected k

**Review by LeonardBesseau — APPROVED** [review]
Good for merging

**Pull request merged**